### PR TITLE
Runtime executor documentation

### PR DESCRIPTION
**Pull request by jwitz**
Resolves https://github.com/astronomer/docs/issues/858

Hey all, in the interest of getting info about the Runtime executor out there in the world, we're publishing some docs under the umbrella of "early access". I know it's been a while since we discussed the executor last, so please let me know what's changed since then and if I'm missing any details (**especially about observability, additional configurations, etc.**

@kwcanuck This is not quite ready for copy review yet

### CODE REVIEW

File: astro/runtime-executor.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+---
+sidebar_label: "Runtime executor"
+title: "Run tasks with the Runtime executor"
+id: runtime-executor
+description: Learn how to your Airflow environment and tasks to take advantage of the Runtime executor, available exclusively in Astro Runtime.
```
Comment (kwcanuck) [inline astro/runtime-executor.md:5]:
```suggestion
description: Learn how to use your Airflow environment and tasks to take advantage of the Runtime executor, available exclusively in Astro Runtime.
```
Alternative description:

Learn how to use the Runtime executor in Astro Runtime.

File: astro/runtime-executor.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+---
+
+The Runtime executor is Runtime's component for executing Airflow tasks. It provides the same functionality as the Celery executor, as well as new features and configurations which make task execution more reliable and observable. See [Runtime image architecture](runtime-architecture.md#runtime-executor) for more information about the benefits of the Runtime executor and how it works.
+
+The Runtime executor can work as a direct replacement for the Celery executor without any additional configuration. You can also use Runtime executor to control task execution to a greater degree than with the Celery executor. For example, you can:
+
+- Specify resource requests for CPU, memory, and concurrency slots within your tasks.
+- Set environment-level strategies how resources should be allocated between tasks that don't have specific resource requests.
```
Comment (kwcanuck) [inline astro/runtime-executor.md:13]:
```suggestion
- Set environment-level strategies to define how resources are allocated between tasks without specific resource requests.
```

File: astro/runtime-executor.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+This functionality is early access and under active development. If you have any questions or feedback about this feature, contact [Astronomer support](https://support.astronomer.io/).
+
+:::
+
+## Enable the Runtime executor in a Deployment
+
+To enable the Runtime executor in a Deployment, set an environment variable with a key of `ASTRO_RUNTIME_EXECUTOR` and a value of `True`. See [Environment variables](environment-variables.md).
+
+After you set this variable, all tasks in the Deployment will run with the Runtime executor. To revert back to using the Celery executor, delete the environment variable from your Deployment.
```
Comment (kwcanuck) [inline astro/runtime-executor.md:26]:
```suggestion
After you set this variable, all tasks in the Deployment will run with the Runtime executor. To revert to the Celery executor, delete the environment variable from your Deployment. 
```

File: astro/runtime-image-architecture.md [hunk 4]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+### Runtime executor
+
+The Runtime executor is exclusive to Astro Runtime and is optimized for resilience and observability.
+
+This executor is a direct replacement for the local and Celery executors. Any DAGs that run using the Celery executor can also run using the Runtime executor.
+
+Additionally you can use the Runtime executor to:
+
+- Specify resource requests for CPU, memory, and concurrency slots within your tasks.
+- Set system-level strategies how resources should be allocated between tasks that don't have specific resource requests.
```
Comment (kwcanuck) [inline astro/runtime-image-architecture.md:102]:
```suggestion
- Set environment-level strategies to define how resources are allocated between tasks without specific resource requests.
```

File: astro/runtime-image-architecture.md [hunk 5]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+
+- The **scheduler** logs the task instance and its allocation requests for newly scheduled tasks in the **Airflow metadata DB**.
+- Allocation is a wrapper around a task instance that indicates its resource requirements, it’s created after the task instance
+- The **Airflow metadata DB** stores both assigned and unassigned allocation requests. The Runtime executor’s DB tables are built in complete parallel and isolation to existing tables. This means that no open source tables in Airflow need to be modified in order for the executor to work.
+- The **allocator** queries the metadata DB for unassigned allocation requests. It then determines how the allocation requests should be handled and submits assignments for the runners back to the DB.
+- **Runners** run tasks. To start running a task, a runner queries the metadata DB to look for new assignment. If it’s been assigned a task, the runner starts completing the work. Currently, there is only one type of runner called a **SubprocessRunner**, which runs tasks within the same cluster as your Airflow environment.
```
Comment (kwcanuck) [inline astro/runtime-image-architecture.md:115]:
```suggestion
- **Runners** run tasks. To start running a task, a runner queries the metadata DB to look for a new assignment. If it’s been assigned a task, the runner starts completing the work. Currently, there is only one type of runner named **SubprocessRunner**, which runs tasks within the same cluster as your Airflow environment.
```

File: astro/runtime-image-architecture.md [hunk 6]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+
+![Runtime executor architecture](/img/docs/runtime-executor-architecture.png)
+
+After a task is scheduled:
+
+- The scheduler logs information about the task and its configured resource requests in Airflow's metadata DB. The task's allocation status is "Unassigned".
```
Comment (kwcanuck) [inline astro/runtime-image-architecture.md:121]:
```suggestion
- The scheduler logs information about the task and its configured resource requests in the Airflow metadata database. The task's allocation status is "Unassigned".
```

File: astro/runtime-image-architecture.md [hunk 7]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+- The allocator looks for any unassigned tasks in the metadata db. When it finds an unassigned task, the allocator determines how to best run the task. It calculates this based on the tasks own resource requests, the configured system-level allocation strategies
```
Comment (kwcanuck) [inline astro/runtime-image-architecture.md:122]:
```suggestion
- The allocator looks for any unassigned tasks in the Airflow metadata database. When it finds an unassigned task, the allocator determines how to best run the task by reviewing existing task resource requests and configured system-level allocation strategies.
```

File: astro/runtime-image-architecture.md [hunk 8]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+- The allocator choses a runner and assigns the task to the runner for execution. After the task is assigned, its allocation status in the metadata db is updated to "Assigned".
```
Comment (kwcanuck) [inline astro/runtime-image-architecture.md:123]:
```suggestion
- The allocator chooses a runner and assigns the task to the runner for execution. After the task is assigned, its allocation status in the Airflow metadata database updates to "Assigned".
```

**Review by kwcanuck — APPROVED** [review]

File: astro/runtime-image-architecture.md [hunk 9]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+- View metrics for task resource usage.
+
+See [Run tasks with the Runtime executor](runtime-executor.md) for setup steps and usage.
+
+### Runtime executor architecture and execution flow
+
+The Runtime executor has a similar execution framework to the Celery executor, but it relies on a few new and existing Airflow components to run tasks:
+
+- The **scheduler** logs the task instance and its allocation requests for newly scheduled tasks in the **Airflow metadata DB**.
+- Allocation is a wrapper around a task instance that indicates its resource requirements, it’s created after the task instance
```
Comment (ianbuss) [inline astro/runtime-image-architecture.md:112]:
```suggestion
- An allocation is a wrapper around a task instance that indicates its resource requirements. It’s created alongside the task instance as a new object in the database.
```

File: astro/runtime-image-architecture.md [hunk 10]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+- The **Airflow metadata DB** stores both assigned and unassigned allocation requests. The Runtime executor’s DB tables are built in complete parallel and isolation to existing tables. This means that no open source tables in Airflow need to be modified in order for the executor to work.
```
Comment (ianbuss) [inline astro/runtime-image-architecture.md:113]:
```suggestion
- The **Airflow metadata DB** stores both assigned and unassigned allocation requests. The Runtime executor’s DB tables are created alongside, but in complete isolation from, existing Airflow tables. This means that no open source tables in Airflow need to be modified in order for the executor to work.
```

File: astro/runtime-image-architecture.md [hunk 11]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
...
+- The **allocator** queries the metadata DB for unassigned allocation requests. It then determines how the allocation requests should be handled and submits assignments for the runners back to the DB.
```
Comment (ianbuss) [inline astro/runtime-image-architecture.md:114]:
```suggestion
- A new component, the **allocator**, queries the metadata DB for unassigned allocation requests. It then determines how the allocation requests should be handled and submits assignments for the runners back to the DB.
```

**Review by ianbuss — COMMENTED** [review]
Looks nice! Left a few comments @jwitz . Am wondering if we need to mention how logging is handled (or at least say you can access task logs for running and completed tasks via the webserver just as you would for Celery)

File: astro/runtime-executor.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+## Enable the Runtime executor in a Deployment
+
+To enable the Runtime executor in a Deployment, set an environment variable with a key of `ASTRO_RUNTIME_EXECUTOR` and a value of `True`. See [Environment variables](environment-variables.md).
+
+After you set this variable, all tasks in the Deployment will run with the Runtime executor. To revert back to using the Celery executor, delete the environment variable from your Deployment. 
+
+## Set task-level resource requests
+
+You can set task-level requests for the following resources:
```
Comment (paolaperaza) [inline astro/runtime-executor.md:30]: Let's give a (brief) explanation here of why you would do this (reliability, especially for bigger tasks)

File: astro/runtime-executor.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+- `cpu`: The number of CPU cores to reserve (for example, `1.5`).
```
Comment (paolaperaza) [inline astro/runtime-executor.md:32]:
```suggestion
- `cpu`: The number of CPU cores to reserve. For example, `1.5`.
```

File: astro/runtime-executor.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+- `memory`: The number of megabytes of RAM to reserve (for example, `1024MB`).
```
Comment (paolaperaza) [inline astro/runtime-executor.md:33]:
```suggestion
- `memory`: The number of megabytes of RAM to reserve. For example, `1024MB`.
```

Is it with `MB` here?

File: astro/runtime-executor.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+- `slots`: The mechanisms for limiting concurrency within a runner. You can request up to 16 slots for a task.
```
Comment (paolaperaza) [inline astro/runtime-executor.md:34]:
```suggestion
- `slots`: The number of tasks that a single task counts for, in the context of controlling worker concurrency, which represents how many tasks a single runner can run at once. You can request up to 64 slots for a single task, which is the concurrency limit for a worker queue on Astro.
```

File: astro/runtime-executor.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+The Runtime allocator uses these requests to determine the best runner for a task to run on. After a task is assigned to a runner, the runner reserves the requested resource for the entire duration of the task. 
+
+You should set these resource requests to be the maximum amount of resources that a task should use at any given time. A task can use some, all, or none of its requested resources. Runners can't scale down the requested resources once the task starts running.
+
+Set task-level resource requests in the task definition. In the following example, a task is configured with a resource request of 1.5 CPU cores and 1024MB.
```
Comment (paolaperaza) [inline astro/runtime-executor.md:40]: This transition feels a little off -- could split it into two sections?

File: astro/runtime-executor.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+            resources= {"cpu": "1.5", "memory": "1024MB", "slots": "2"},
+        )
+
+```
+
+When this task is assigned to the runner, the runner reserves 1.5 CPU cores and 1024MB for the entirety of the task, even if the task only uses 512MB of memory.
+
+### How slot requests work
+
+Slots are a mechanism for limiting the amount of tasks that can run concurrently on a single runner. By default, runners have 16 slots and tasks request 1 slot.
```
Comment (paolaperaza) [inline astro/runtime-executor.md:55]: I think we should actually split up into two sub-sections:

## Task-level Requests
[high-level overview of two methods and why you'd use them]

### CPU and Memory

### Slots

File: astro/runtime-executor.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+You might want to request more slots for large tasks that you don't want limited by specific resource requests for CPU and memory. Running a task in multiple slots ensures that the task doesn't compete for resources with other tasks in the same runner.
+
+## Set environment-level allocation strategies
```
Comment (paolaperaza) [inline astro/runtime-executor.md:59]:
```suggestion
## Set an allocation strategy
```

Don't think you ever set more than "one" or plural strategies

File: astro/runtime-executor.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+Set the `ALLOCATOR_STRATEGY` [environment variable](environment-variables.md) to configure how the Runtime executor allocates tasks across your runners. This environment variable has two possible values:
+
+- `EvenSpreadStrategy`: The allocator attempts to assign tasks evenly across all available runners.
+- `MaxPackingStrategy`: The allocator attempts to assign as many tasks as possible to a single runner before using another runner.
```
[the remark this answers is not in the document]
Comment (paolaperaza) [reply astro/runtime-executor.md:64]: If we want to, I'd suggest we put it under some section called "Advanced" or something

File: astro/runtime-executor.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+### How slot requests work
+
+Slots are a mechanism for limiting the amount of tasks that can run concurrently on a single runner. By default, runners have 16 slots and tasks request 1 slot.
+
+You might want to request more slots for large tasks that you don't want limited by specific resource requests for CPU and memory. Running a task in multiple slots ensures that the task doesn't compete for resources with other tasks in the same runner.
+
+## Set environment-level allocation strategies
+
+Set the `ALLOCATOR_STRATEGY` [environment variable](environment-variables.md) to configure how the Runtime executor allocates tasks across your runners. This environment variable has two possible values:
```
Comment (paolaperaza) [inline astro/runtime-executor.md:61]: @jwitz Can we add a section on how to view runner status in the Airflow UI? Do you have access to that? @ianbuss Would be good to get more guidance here on the Runtime team on what's available in that runner tab and how we expect users to reference it

File: astro/runtime-image-architecture.md [hunk 21]
```diff
@@ -88,9 +88,40 @@ Astro Runtime supports Python 3.9. This is the only version of Python that Astro
 
 In Airflow, the executor is responsible for determining how and where a task is completed.
 
-In all local environments created with the Astro CLI, Astro Runtime runs the [Local executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/local.html). On Astro, Astro Runtime exclusively supports the [Celery executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html).
+In all local environments created with the Astro CLI, Astro Runtime runs the [Local executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/local.html). On Astro, Astro Runtime supports both the [Celery executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) and the Astronomer-developed Runtime executor.
 
-Soon, Astronomer will provide a new executor with intelligent worker packing, task-level resource requests, improved logging, and Kubernetes-like task isolation.
+### Runtime executor
```
Comment (paolaperaza) [inline astro/runtime-image-architecture.md:93]: In reading this again, I'm finding myself referencing this architecture doc and the enablement doc back-and-forth to understand "how it works and how I should use it."

I'm really thinking we just need to consolidate all of this into the runtime executor doc at the top (dedicated architecture section) and then keep the Runtime IMAGE architecture doc higher level about which executors it supports and then link out

**Review by paolaperaza — CHANGES_REQUESTED** [review]

File: astro/runtime-executor.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
+---
+sidebar_label: "Runtime executor"
+title: "Run tasks with the Runtime executor"
```
[the remark this answers is not in the document]
Comment (jwitz) [reply astro/runtime-executor.md:3]: Cut some words here, keeping things focused on "Resources"

File: astro/runtime-executor.md (hunk 20 above)
Comment (jwitz) [reply astro/runtime-executor.md]: I've added this section, but I'm running into a few bugs that made me doubt whether we should be writing about this now - even if it's still in beta.

**Comment by jwitz** [comment]
@paolaperaza This is ready for review again ✅ Made quite a few changes this time around but mainly stuck to the current suggestions. There's also a new section about the **Executor Status** page in the Airflow UI.

File: astro/runtime-executor.md [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+#### Resource request example
+
+The following task is configured with a resource request of 1.5 CPU cores and 1024MB.
+
+```python
+my_task = BranchPythonOperator(
+            task_id="my_task",
+            python_callable=_get_activity, # Python function called when task executes
+            resources= {"cpu": "1.5", "memory": "1024MB", "slots": "2"},
```
Comment (jarfgit) [inline astro/runtime-executor.md:95]: Syntax for the resource request is just slightly off here. It should be:
`executor_config={"resources": {"cpu": "1.5", "memory": "1024MB", "slots": "2"}},`

File: astro/runtime-executor.md [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+flowchart LR;
+classDef astro fill:#dbcdf6,stroke:#333,stroke-width:2px;
+    A[Scheduler]-->|Log new tasks/ allocation requests| B[(Metadata DB)];
+    C[Allocator]-->|Query for new allocations / Log new assignments| B;
+    D[Runners]-->|Query for new assignments/ log status|B;
+    E[Webserver]-->|Query for results of runners' work|B;
+    style ControlPlane fill:#bfeaff,stroke:#333,stroke-width:2px
+    style DataPlane fill:#bfeaff,stroke:#333,stroke-width:2px
+    style AstroCLI fill:#bfeaff,stroke:#333,stroke-width:2px
+    style Deployment fill:#bfeaff,stroke:#333,stroke-width:2px
```
Comment (jwitz) [inline astro/runtime-executor.md:46]:
```suggestion
```

**Comment by paolaperaza** [comment]
The Runtime Executor is unfortunately not imminent and still unplanned, at this point. Closing for now and we'll resurrect if/when needed 😞

**Pull request closed without merging**